src/modules/image.cxx

```diff
@@ -0,0 +1,130 @@
+//use lodepng, and (maybe do smth about the alpha channel)
+module;
+
+#include "lodepng.h"
+#include <vector>
+#include <iostream>
+#include <cstdint>
+#include <string>
+#include <filesystem>
+#include <fstream>
+
+export module Image;
+
+export class Pixel
+{
+
+    public:
+    uint8_t r, g, b;
+
+    Pixel() {}
+    Pixel(uint8_t r, uint8_t g, uint8_t b) : r(r), g(g), b(b) {}
+
+    friend std::istream& operator>>(std::istream& is, Pixel& pixel)
+    {
+        is >> pixel.r >> pixel.g >> pixel.b;
+        return is;
+    }
+    friend std::ostream& operator<<(std::ostream& os, Pixel& pixel)
+    {
+        os << std::to_string(pixel.r) << " " << std::to_string(pixel.g) << " " << std::to_string(pixel.b);
+        return os;
+    }
+    friend std::ostream& operator<<(std::ostream& os, const Pixel& pixel)
+    {
+        os << std::to_string(pixel.r) << " " << std::to_string(pixel.g) << " " << std::to_string(pixel.b);
+        return os;
+    }
+};
+
+
+export class Image
+{
+public:
+    std::vector<Pixel> pixels;
+    const uint8_t  max_value;
+    unsigned int width, height;
+
+    Image(unsigned int width, unsigned int height, uint8_t max_value, const std::vector<Pixel>& pixels) : width(width), height(height), max_value(max_value), pixels(pixels) {}
+
+    Image(unsigned int width, unsigned int height, uint8_t max_value, const std::vector<uint8_t>& pixels) : width(width), height(height), max_value(max_value) {
+        this->pixels.reserve(pixels.size() / 3);
+        for (int i = 0; i < pixels.size(); i += 3)
+            this->pixels.push_back(Pixel(pixels[i], pixels[i + 1], pixels[i + 2]));
+    
+    }
+
+    Image(int width, int height, int max_value) : width(width), height(height), max_value(max_value) {}
+
+    unsigned int&
+    get_width()
+    {
+        return width;
+    }
+
+    unsigned int&
+    get_height()
+    {
+        return height;
+    }
+
+    void
+    read_png(std::string filename)
+    {
+        std::vector<unsigned char> image;
+        unsigned int w, h;
+        unsigned int error = lodepng::decode(image, w, h, filename, LCT_RGB, 8);
+        if (error)
+            std::cout << "decoder error " << error << ": " << lodepng_error_text(error) << std::endl;
```

Comment:
When writing a library (as we do here), direct error handling should be avoided. What if the caller of this function wants to do something else when this error happens?

It is debatable how to handle it, but I'd suggest throwing an exception.